File: frameworks/node/koa/Implementation/routes/parameters.ts
import type { Routes } from "../app.ts";

/**
 * parameters: router captures, each echoed back as an integer. @koa/router runs the matching
 * routes in the order they were registered, so the static path is registered first and answers
 * before the capturing route that also matches it.
 */
const parameters: Routes = (router, { payloads: p }) => {
  // rb:wiring parameters.*
  // A capture reaches the route as the string it matched. router.param is @koa/router's hook for a
  // named capture, which runs once a route with it matches and before that route's middleware, so
  // each capture is converted there and kept on ctx.state, where Koa passes values between middleware.
  router.param("one", (value, ctx, next) => {
    ctx.state.one = Number(value);
    return next();
  });
  router.param("two", (value, ctx, next) => {
    ctx.state.two = Number(value);
    return next();
  });
  // rb:end

  router.get("/parameters/static/segment/literal", (ctx) => {
    ctx.body = p.small;
  });

  router.get("/parameters/:one/segment/literal", (ctx) => {
    ctx.body = { ...p.small, echo: { one: ctx.state.one } };
  });

  router.get("/parameters/:one/with-second/:two", (ctx) => {
    ctx.body = { ...p.small, echo: { one: ctx.state.one, two: ctx.state.two } };
  });
};

export default parameters;

File: frameworks/java/ktor/Implementation/src/main/kotlin/implementation/routes/Parameters.kt
package implementation.routes

import implementation.Payloads
import implementation.echoed
import io.ktor.server.response.respond
import io.ktor.server.routing.Route
import io.ktor.server.routing.get
import io.ktor.server.util.getOrFail
import kotlinx.serialization.Serializable

@Serializable
data class One(val one: Int)

@Serializable
data class Two(val one: Int, val two: Int)

/**
 * parameters: path captures, each converted to an Int by getOrFail, which runs Ktor's data conversion
 * and refuses a value that is missing or does not convert with 400. Ktor's router tries a constant
 * segment before a capture.
 */
fun Route.parameters(p: Payloads) {
    get("/parameters/static/segment/literal") { call.respond(p.small) }
    get("/parameters/{one}/segment/literal") {
        call.respond(p.small.echoed(One(call.parameters.getOrFail<Int>("one"))))
    }
    get("/parameters/{one}/with-second/{two}") {
        call.respond(p.small.echoed(Two(call.parameters.getOrFail<Int>("one"), call.parameters.getOrFail<Int>("two"))))
    }
}
